harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +35,143 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def _merge_metadata(
+    merged: Dict[str, Any],
+    rank_metadata: Dict[str, Any],
+    rank: int,
+    key_ranks: Dict[str, Any],
+    key_conflicts: Dict[str, List[int]],
+    prev_key: str,
+) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+
+    for key, metadata in rank_metadata.items():
+
+        # First, update key_ranks. We will use it later
+        if key not in key_ranks:
+            key_ranks[key] = {"_ranks_": []}
+        key_ranks[key]["_ranks_"].append(rank)
+
+        full_key = f"{prev_key}/{key}"
+
+        # Merge metadata. We consider two possibilities:
+        # 1. metadata is dictionary AND :
+        #   * if key is not in merged, then recursively unroll and merge metadata; OR
+        #   * if key is in merged, and metadata and merged[key] are dictionaries, merge recursively.
+        # 2. otherwise:
+        #   * key is in merged, then updated key_conflicts and don't merge,
+        #   * key is not in merged, then add key and metadata to merged.
+        if (key not in merged and isinstance(metadata, dict)) or (
+            key in merged and isinstance(metadata, dict) and isinstance(merged[key], dict)
+        ):
+            merged[key], key_ranks[key] = _merge_metadata(
+                {} if key not in merged else merged[key],
+                metadata,
+                rank,
+                key_ranks[key],
+                key_conflicts,
+                full_key,
+            )
+        else:
+            if key in merged and merged[key] != metadata:
+                if full_key not in key_conflicts:
+                    key_conflicts[full_key] = key_ranks[key]["_ranks_"].copy()
+                else:
+                    key_conflicts[full_key].append(rank)
+            else:
+                merged[key] = metadata
+
+    return merged, key_ranks
+
+
+def merge_metadata(
+    all_metadata: List[Dict[str, Any]]
+) -> Tuple[Dict[str, Any], Dict[str, List[int]]]:
+    """
+    Given a list of metadata, return:
+      - merged metadata
+      - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+      - merge dictionaries under the same key
+      - report conflicts for any repeated key that is not a dictionary
+
+    Examples:
+    #  Metadata 1     | Metadata 2       | Conflict |  Result
+    # ----------------|------------------|----------|-----------------
+    #  a: 1           | b: 2             | no       | {a: 1, b: 2}
+    #  a: 1           | a: 1             | no       | {a: 1}
+    #  a: []          | a: []            | yes      | n/a
```

Comment:
would this example result in a conflict after changing to allow overlapped keys if values are the same?

Reply:
values are compared with `==` which for two lists means that as long as they have the same content, there is no conflict. I will update the description. thanks for catching it!

as pointed by Ryan:
> there is actually a [published RFC about json merging](https://www.rfc-editor.org/rfc/rfc7386.html), and they don't allow merging lists

and we decided to follow that and treat lists as any other values since there is no standardized way of merging.